Discussion task overview (#58)
ALLUOS/ALLUOS.github.io: _posts/2021-03-02-discussion.md

Final feedback on users' participation closes the discussion task.

### Theoretical framework

A great number of existing studies have appraised the role played by discussion tasks in second/foreign language acquisition. In the introduction to this section, we explained the motivations and intentions that led us to develop a discussion task. We wanted to build a space for our users to interact more freely among each other, and prior research suggests that group discussion relaxes the stress on the learners because the focus is not on getting a specific target exercise right, but rather on communicating more generally (Puspitasari, 2016). Naturally, a number of specific design choices had to be made when deciding how to tailor a discussion task to our specific learning environment. In this section we will review the literature that informed each of these choices.

> As already noted, we wanted this task to enhance language production skills. Discussions are a staple of classroom teaching, not only in the context of second language learning. They provide opportunities to produce language in contextualized and purposeful ways: learners can practice applying form (e.g., grammar, vocabulary) and function (e.g., language used to clarify, explain, argue) to communicate and build ideas. Moreover, learners benefit from redundancy of ideas and the vocabulary associated with them: discussion allows for multiple opportunities to hear new concepts and content continuously explained, analyzed, and interpreted. Furthermore, discussions via written communication have been linked to the immersive experience of studying abroad (Oliva & Pollastrini, 1995), notoriously effective for language learning.

Reviewer: ```suggestion
As already noted, we wanted this task to enhance language production skills. Discussions are a staple of classroom teaching, not only in the context of second language learning. They provide opportunities to produce language in contextualized and purposeful ways: learners can practice applying form (e.g., grammar, vocabulary) and function (e.g., language used to clarify, explain, argue) to communicate and build ideas. Moreover, learners benefit from redundancy of ideas and the vocabulary associated with them: discussion allows for multiple opportunities to hear new concepts and content continuously explained, analyzed, and interpreted. Furthermore, discussions via written communication have been likened to the immersive experience of studying abroad (Oliva & Pollastrini, 1995), notoriously effective for language learning.
```

### Outlook

We think that the discussion task is a nice addition to the first two tasks. This could even be further developed in the future.

Possible future extensions of the task can include messages from the bot to discuss with users. An advantage of these contributions to the discussion can be more guidance and the effect of “model answers” which can serve for less experienced English speakers as an orientation. This also includes examples for the use of discussion phrases which can add to the learning effect.

Another additional feature can be corrections and helpful hints by the bot. Meanwhile, the bot should be similar as Barson (1993) describes the teacher in a similar classroom setting: only play a supportive role and help discreetly to identify errors while mostly appreciating the learner’s contributions.

Furthermore, vocabulary hints and short descriptions of words which are possibly new to the learners would be a valuable addition. This way probably an even bigger vocabulary learning effect could be achieved.

## Implementation process

> Below, the technical side of the discussion task reviewed. As part of our overall bot architecture, the implementation of the discussion task also relies on general modules like the `Task(ABC)` class and the room handler, described in the [implementation section]({{ "" | absolute_url }}/2021/04/02/implementation.html). Therefore, only the discussion-specific task handler is described here in detail.

Reviewer: ```suggestion
Here, we present the technical details of the implementation of this task. As part of our overall bot architecture, the implementation of the discussion task also relies on general modules like the `Task(ABC)` class and the room handler, described in the [implementation section]({{ "" | absolute_url }}/2021/04/02/implementation.html). Therefore, only the discussion-specific task handler is described here in detail.
```

`Discussion(Task)` is a subclass of `Task(ABC)`. However, unlike the sentence correction and the vocabulary guessing tasks, the discussion task is not sequential in nature and, therefore, does not inherit from `SequentialTask`. Nevertheless, we were still able to reuse a significant part of the methods already implemented in said class.
The task always starts off with the presentation of the text, which is retrieved from the database based on the users' proficiency. This is followed by three rounds of discussion.

> At the start of each round, a text-related question is sent and a timer is added to the job queue to determine when the round finishes, such that [adaptive graphics]({{ "" | absolute_url }}/2021/01/04/user/experience.html) can remind users to finish their discussion of the question on time.

Reviewer: ```suggestion
At the start of each round, a text-related question is sent and a timer is added to the job queue to determine when the round finishes, such that [visual reminders]({{ "" | absolute_url }}/2021/02/03/adaptive.html) can be sent at regular time intervals to remind users to finish their discussion of the question on time.
```

> Since we don't have to select for a specific user, all user input can be handled in the same `handle_question` state. Word counts and grammar scores for each user are stored in two dictionaries, which are required to satisfy the completion criteria of the task.

Reviewer: ```suggestion
Unlike other [tasks]({{ "" | absolute_url }}/#three), in this case we don't have to select a specific user, therefore all user input can be handled in the same `handle_question` state. Word counts and grammar scores for each user are stored in two dictionaries, which are required to satisfy the completion criteria of the task.
```

> After a message is sent, `update_word_counts(message, user)` and `grammar_classifier.classify(sentence)` are called in the `evaluate_user_input(update.message.from_user.username, update.message.text, task_no=1)` function. They are called on tokenized and preprocessed input in order to update those dictionaries with the information.

Reviewer: ```suggestion
After a message is sent, `update_word_counts(message, user)` and `grammar_classifier.classify(sentence)` are called in the `evaluate_user_input(update.message.from_user.username, update.message.text, task_no=1)` function. They are called on tokenized and preprocessed input in order to update those dictionaries with the relevant information.
```

In the first and the second discussion rounds, additionally to evaluating user input, the `send_intermediate_feedback()` function is called.

> In the third and final discussion round, instead of intermediate feedback the final grammatical feedback is presented via calling `send_final_feedback()`.

Reviewer: I noticed that the mention of feedback comes up here for the first time, without a previous explanation. Even though there are several pointers to the GEC page, I'm afraid that someone who's only read this page so far might be confused by this. We might want to add a sentence or two about this earlier in the "completion metrics" section :)
